### PR TITLE
Implement rectangle area support for some support powers.

### PR DESCRIPTION
**Pull request by GraionDilach**
The first commit is actually from #11624, which IMO is the longterm solution for resolving https://github.com/OpenRA/OpenRA/pull/14889#discussion_r176625652.

### CODE REVIEW

**Comment by GraionDilach** [comment]
Note that RA2 will need this since it applied these powers in a 3x3 rectangle.

**Comment by pchote** [comment]
Would you mind changing this to support arbitrary footprints instead (see e.g. Building, TerrainTunnel)?  This would be a lot more flexible, and should lead to simpler code too (no conditionals).

A 3x3 rectangle would be:
```yaml
Dimensions: 3, 3
Footprint: xxx xxx xxx
```
and a 3x3 circle would be 
```yaml
Dimensions: 3, 3
Footprint: _x_ xxx _x_
```
and so on for any other shapes.

**Comment by GraionDilach** [comment]
TBH, I'd like to see the first commit merged in, because RA2-accurate cluster weapons also used a 3x3 area and I wouldn't do that with foundation setup, but I presume I could do that.... ohwait, that needs a new-generation upgrade rule.

Bleh.

**Comment by GraionDilach** [comment]
Lost interest in this one.

**Comment by GraionDilach** [comment]
Since my prelimary tests were interesting in regards to footprints and I actually have time to focus on development again, I am reopening this.

**Comment by pchote** [comment]
Re-requesting my footprint comment / example from above.

**Comment by GraionDilach** [comment]
Sure.

**Comment by GraionDilach** [comment]
Partially completed now, still working on the upgrade rule.

File: OpenRA.Mods.Common/UpdateRules/Rules/ConvertSupportPowerRangesToFootprint.cs [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+			power.AddNode(new MiniYamlNode("Dimensions", (size).ToString() + ", " + (size).ToString()));
+
+			string footprint = "";
+
+			int emptycell = range, affectedcell = 1;
+
+			for (int i = 0; i < size; i++)
+			{
+				for (int e = 0; e < emptycell; e++)
+				{
```
Comment (reaperrr) [inline OpenRA.Mods.Common/UpdateRules/Rules/ConvertSupportPowerRangesToFootprint.cs]: Style nit (you can wait for more feedback/change requests before fixing this): we only use curly brackets around single-line statements if necessary for readability (which isn't the case here).

**Comment by GraionDilach** [comment]
Fixed typo and removed unnecessary braces.

**Comment by GraionDilach** [comment]
This needs a rebase, but I won't do it until there is atleast one review on it.

**Comment by GraionDilach** [comment]
Rebased to fix a braces error.

**Comment by abcdefg30** [comment]
Seems like this doesn't compile now.

**Comment by GraionDilach** [comment]
Yeah, I ended up confusing myself. >_> Meh, added the newlines to prevent overlooking the necessity of the braces inb4 I pull it off again.

**Comment by pchote** [comment]
Moving to the Future milestone so we don't lose this.

**Comment by GraionDilach** [comment]
Depends on #15415 now.

**Comment by GraionDilach** [comment]
Rebased.

File: OpenRA.Mods.Common/OpenRA.Mods.Common.csproj [hunk 2]
```diff
@@ -945,4 +946,4 @@
     <Message Text="DEBUG TargetName:             $(TargetName)"/>
     -->
   </Target>
-</Project>
\ No newline at end of file
+</Project>
```
Comment (pchote) [inline OpenRA.Mods.Common/OpenRA.Mods.Common.csproj]: Newline.

File: OpenRA.Mods.Common/UpdateRules/Rules/ConvertSupportPowerRangesToFootprint.cs [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+			if (rangeNode != null)
+			{
+				range = rangeNode.NodeValue<int>();
+				power.RemoveNode(rangeNode);
+			}
+
+			var size = 2 * range + 1;
+			power.AddNode(new MiniYamlNode("Dimensions", size.ToString() + ", " + size.ToString()));
+
+			string footprint = string.Empty;
```
Comment (pchote) [inline OpenRA.Mods.Common/UpdateRules/Rules/ConvertSupportPowerRangesToFootprint.cs]: `var` here and in a bunch more places below.

**Review by teinarss — DISMISSED** [review]
* Tested the changes
* Tested the UpdateRules
No comments about the code.

**Comment by GraionDilach** [comment]
Fixed.

**Comment by atlimit8** [comment]
Superseded by #17786. Removing from milestone.

**Pull request closed without merging**